Template Duplication

Daniel-Jang opened the pull request:
Original Issue: https://github.com/cass-degrees/CASS-Degrees-Code/issues/18

This allows the user to duplicate an existing degree/program that a user has selected from list.html for the user to alter it and create.

Daniel-Jang committed:
Allow the user to click on "Create From" button in list.html, which opens the create program template with pre-filled form with data corresponding to the ID the user chose to duplicate from.

pep8speaks commented:
Hello @Daniel-Jang! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`cassdegrees/ui/views/programs.py`](https://github.com/cass-degrees/CASS-Degrees-Code/blob/e193fa76d8c1d8ed60c8089ce40d7c9ce67ee543/cassdegrees/ui/views/programs.py):

> [Line 11:9](https://github.com/cass-degrees/CASS-Degrees-Code/blob/e193fa76d8c1d8ed60c8089ce40d7c9ce67ee543/cassdegrees/ui/views/programs.py#L11): [F841](https://duckduckgo.com/?q=pep8%20F841) local variable 'create_from' is assigned to but never used

Daniel-Jang committed:
Altered the name of feature "Create from" to "Duplicate" as it is a better choice of phrase.

Daniel-Jang committed:
Corrected some error I have forgotten to change.

j-selby commented on cassdegrees/ui/views/programs.py:
```diff
@@ -14,7 +30,10 @@ def create_program(request):
 
     else:
-        form = EditProgramFormSnippet()
+        if duplicate:
```
Wait, this will edit, not duplicate. You need to reset the ID, no?

Daniel-Jang replied:
No, I tried it out and it did not edit the existing one. That part of code is for GET requests, and that instance=instance was done just to display the contents of the program user selected to duplicate from.

I have left the POST part of create_program view function as it was before so adding is normal. Attempting to add a duplicate without changing anything will lead to this anyways:
![image](https://user-images.githubusercontent.com/37033052/56935276-4fedcf00-6b33-11e9-8b9c-df3589de54dd.png)

j-selby commented on cassdegrees/templates/list.html:
```diff
@@ -66,6 +66,10 @@
                                         <td>
+                                            {# Only Degree and Subplans will need this feature #}
+                                            {% if title == 'Degree' or title == 'Subplan'  %}
+                                                <a class="btn-uni-grad" href="/create/{% if title == 'Degree' %}program{% elif title == 'Subplan' %}subplan{% elif title == 'Course' %}course{% endif %}/?id={{ item }}&duplicate=True">Duplicate</a>
```
Web syntax is generally lower-case false here.

j-selby commented on cassdegrees/ui/views/programs.py:
```diff
@@ -6,6 +6,22 @@
 
 
 def create_program(request):
+    duplicate = request.GET.get('duplicate', False)
```
You are either returning a string or a boolean value here - little inconsistent.

Daniel-Jang replied:
Fixed it by setting it to 'false' string and then reassign duplicate variable as boolean True or False immediately after receiving it.

Daniel-Jang committed:
Made duplication variable assignment more consistent and set 'True' to 'true' in list.html for the duplication link, as lower case seems to be the web syntax.

Daniel-Jang committed:
Merge branches 'develop' and 'feature-template_duplication' of https://github.com/cass-degrees/CASS-Degrees-Code into feature-template_duplication

Daniel-Jang committed:
Since CoursesInSubplanModel no longer exists due to Aiden's pull request, I simply removed it from the import list.
Everything seems to work fine anyways so I will leave it like this (otherwise it crashes).

Daniel-Jang committed:
Merge branches 'develop' and 'feature-template_duplication' of https://github.com/cass-degrees/CASS-Degrees-Code into feature-template_duplication

Daniel-Jang committed:
Merge branches 'develop' and 'feature-template_duplication' of https://github.com/cass-degrees/CASS-Degrees-Code into feature-template_duplication

# Conflicts:
#	cassdegrees/templates/list.html

Daniel-Jang committed:
Removed redundant elif in list.html and fixed issues caused by the Degree name change to Programs

JackSekoranja reviewed (approved):
Looks good, works well. No complaints

KavsHub reviewed (approved):
This is pretty much the same as the subplan one. Happy for you to merge both of them in.
